torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -609,9 +663,14 @@ def _should_free_in_backward(
     Returns whether FSDP should free the unsharded flattened parameter in the
     post-backward or not.
     """
-    return (
-        state._sync_gradients and handle.uses_sharded_strategy
-    ) or handle._config.sharding_strategy == HandleShardingStrategy.FULL_SHARD
+    # We always free if we are syncing gradients (i.e. not in no_sync) and parameters
+    # are sharded.
+    free_unsharded = state._sync_gradients and handle.uses_sharded_strategy
+    # For NO_SHARD we don't need to free full parameters, for Zero-2 strategies, we skip
```

Comment:
We should have a discussion about if we want to refer to `reshard_after_forward = False` strategies as ZeRO-2. I am personally not a big fan because ZeRO-2 is not accurate, and it also requires users to learn about ZeRO-2/3 differences to understand our APIs, which will then also lead them to more confusion since they do not align.

Reply:
[Not blocking] However, we should definitely discuss sometime

Reply:
[Not blocking] However, I recommend us think about how to refer to the `reshard_after_forward = False` strategies. The debt will only build with more PRs on top of this.

Reply:
agreed, we should consider this in follow up discussions